Objective-C/Tests/QueryTest+Main.m

```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
+        count++;
+        AssertNotNil(change.query);
+        AssertNil(change.error);
+        NSArray<CBLQueryResult*>* rows = [change.results allObjects];
+        if (count == 1) {
+            AssertEqual(rows.count, 9u);
+        } else {
+            XCTFail(@"Unexpected update from LiveQuery");
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)),
+                   dispatch_get_main_queue(), ^{
+                       // This change will not affect the query results because 'number1 < 10' is not true.
+                       [self createDocNumbered: 111 of: 100];
+                   });
+    
+    // Wait 2 seconds, then fulfil the expectation:
+    XCTestExpectation *x = [self expectationWithDescription: @"Timeout"];
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)),
+                   dispatch_get_main_queue(), ^{
+                       [x fulfill];
+                   });
+    
+    NSLog(@"Waiting...");
+    [self waitForExpectationsWithTimeout: 5.0 handler: ^(NSError *error) { }];
+    NSLog(@"Done!");
+    
+    AssertEqual(count, 1);
+    
+    [q removeChangeListenerWithToken: token];
+}
+
+/**
+ When adding a second listener after the first listener is notified, the second listener
+ should get the change (current result).
+ */
+- (void) testLiveQuerySecondListenerReturnsResultsImmediately {
+    [self createDocNumbered: 7 of: 10];
+    CBLQuery* q = [CBLQueryBuilder select: @[[CBLQuerySelectResult property: @"number1"]]
+                                     from: [CBLQueryDataSource database: self.db]
+                                    where: [[CBLQueryExpression property: @"number1"] lessThan: [CBLQueryExpression integer: 10]]
+                                  orderBy: @[[CBLQueryOrdering property: @"number1"]]];
+    
+    // first listener
+    XCTestExpectation* first = [self expectationWithDescription: @"first-set-of-notifications"];
+    __block int count = 0;
+    id token1 = [q addChangeListener: ^(CBLQueryChange* change) {
+        count++;
+        NSArray<CBLQueryResult*>* rows = [change.results allObjects];
+        if (count == 1) {
+            
+            // initial notification about already existing result set!
+            AssertEqual(rows.count, 1);
+            AssertEqual([rows[0] integerAtIndex: 0], 7);
+        } else if (count == 2) {
+            
+            // newly added doc notification
+            AssertEqual(rows.count, 2);
+            AssertEqual([rows[0] integerAtIndex: 0], 3);
+            
+            [first fulfill];
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        [self createDocNumbered: 3 of: 10];
+    });
+    
+    // wait for first set of notifications
+    [self waitForExpectations: @[first] timeout: 10.0];
+    AssertEqual(count, 2);
+    [q removeChangeListenerWithToken: token1];
```

Comment:
It's better to not remove the token1 and to make sure that after adding the second listener, the first listener doesn't get the notification.